// angular2-tree-component/dist/components/tree-node-content.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var tree_node_model_1 = require('../models/tree-node.model');
var tree_model_1 = require('../models/tree.model');
var TreeNodeContent = (function () {
    function TreeNodeContent(treeModel, componentResolver, viewContainerRef) {
        this.treeModel = treeModel;
        this.componentResolver = componentResolver;
        this.viewContainerRef = viewContainerRef;
    }
    TreeNodeContent.prototype.ngAfterViewInit = function () {
        this._loadTreeNodeContent();
    };
    TreeNodeContent.prototype._loadTreeNodeContent = function () {
        var _this = this;
        this.componentResolver.resolveComponent(this.treeModel.treeNodeContentComponent)
            .then(function (componentFactory) {
            var componentRef = _this.viewContainerRef.createComponent(componentFactory, 0, _this.viewContainerRef.injector);
            componentRef.instance.node = _this.node;
            componentRef.changeDetectorRef.detectChanges();
        });
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', tree_node_model_1.TreeNode)
    ], TreeNodeContent.prototype, "node", void 0);
    TreeNodeContent = __decorate([
        core_1.Component({
            selector: 'TreeNodeContent',
            template: '',
        }), 
        __metadata('design:paramtypes', [tree_model_1.TreeModel, core_1.ComponentResolver, core_1.ViewContainerRef])
    ], TreeNodeContent);
    return TreeNodeContent;
}());
exports.TreeNodeContent = TreeNodeContent;
